thewalrus/_torontonian.py

```diff
@@ -54,3 +54,193 @@ def tor(A, fsum=False):
+    Yields:
+        results (list) : length r combination from pool
+    """
+    n = len(pool)
+    indices = list(range(r))
+    empty = not (n and (0 < r <= n))
+
+    if not empty:
+        result = [pool[i] for i in indices]
+        yield result
+
+    while not empty:
+        i = r - 1
+        while i >= 0 and indices[i] == i + n - r:
+            i -= 1
+        if i < 0:
+            empty = True
+        else:
+            indices[i] += 1
+            for j in range(i + 1, r):
+                indices[j] = indices[j - 1] + 1
+
+            result = [pool[i] for i in indices]
+            yield result
+
+
+@numba.jit(nopython=True)
+def powerset(S):
+    """
+    generates the powerset of S
+
+    does not include the empty set 
+
+    Args:
+        S (array/iterable) : set to take powerset from
+    Yields:
+        s (list) : subset of S
+    """
+    n = len(S)
+    for i in range(n + 1):
+        for s in combinations(S, i):
+            yield s
+
+
+@numba.jit(nopython=True)
+def nb_block(X):
+    """ 
+    numba implementation of np.block
+    taken from: https://stackoverflow.com/a/57562911
+
+    Args:
+        X (tuple of arrays) : arrays for blocks of matrix
+    Return:
+        array : the block matrix from X
+    """
+    xtmp1 = np.concatenate(X[0], axis=1)
+    xtmp2 = np.concatenate(X[1], axis=1)
+    return np.concatenate((xtmp1, xtmp2), axis=0)
+
+
+@numba.jit(nopython=True)
+def numba_ix(arr, rows, cols):
+    """
+    numba implementation of np.ix_
+
+    Args:
+        arr (2d array) : matrix to take submatrix of
+        rows (array) : rows to be selected in submatrix
+        cols (array) : columns to be selected in submatrix
+
+    Return: 
+        len(rows) * len(cols) array : selected submatrix of arr
+    """
+    return arr[rows][:, cols]
+
+
+@numba.jit(nopython=True)
+def Qmat_numba(cov, hbar=2):
+    r"""
+    numba compatible version of thewalrus.quantum Qmat
+
+    Returns the :math:`Q` Husimi matrix of the Gaussian state.
+    Args:
+        cov (array): :math:`2N\times 2N xp-` Wigner covariance matrix
+        hbar (float): the value of :math:`\hbar` in the commutation
+            relation :math:`[\x,\p]=i\hbar`.
+    Returns:
+        array: the :math:`Q` matrix.
+    """
+    # number of modes
+    N = len(cov) // 2
+    I = np.identity(N)
+
+    x = cov[:N, :N] * (2. / hbar)
+    xp = cov[:N, N:] * (2. / hbar)
+    p = cov[N:, N:] * (2. / hbar)
+    # the (Hermitian) matrix elements <a_i^\dagger a_j>
+    aidaj = (x + p + 1j * (xp - xp.T) - 2 * I) / 4
+    # the (symmetric) matrix elements <a_i a_j>
+    aiaj = (x - p + 1j * (xp + xp.T)) / 4
+
+    # calculate the covariance matrix sigma_Q appearing in the Q function:
+    # Q(alpha) = exp[-(alpha-beta).sigma_Q^{-1}.(alpha-beta)/2]/|sigma_Q|
+    Q = nb_block(((aidaj, aiaj.conj()), (aiaj, aidaj.conj()))) + np.identity(2 * N)
+    return Q
+
+
+@numba.jit(nopython=True)
+def threshold_detection_prob(mu, cov, det_pattern, hbar=2):

```

Comment:
I'd suggest this function calculate only the numerator of the probability, not the prefactors involving exponentials and the inverse of the square of the determinant of the covariance matrix. Ideally it would have the same signature as the torontonian except that it would take a vector of displacements